### DISCUSSION TITLE
Editable ComboBox

### DISCUSSION THREAD

**Discussion by ManuelKetisch** [AvaloniaUI/Avalonia#13301] [Q&A] [answered]
Hello Avalonia community,

i am currently working on a small application which uses a few TextBoxes and ComboBoxes.
The user is supposed to provide data, which is then further processed.
To make it easier for the user, i want to use ComboBoxes as a way to provide some common entries, the user only needs to select. 
However, it might happen, that the user needs to enter something, the ComboBox does not provide.
Because of that, i would like these ComboBoxes to be editable TextBoxes, so if the user needs to input data that the CB does not provide, he can simply type in the required data by hand.

Is there a way to implement this functionality without the need to code an entirely new control?
Are there any plans on implementing this to base Avalonia as kind of a property like IsEditable?

**Comment by thevortexcloud** [comment]
It looks like you are not the first person to ask this: https://github.com/AvaloniaUI/Avalonia/issues/205

**Reply by timunie** [reply]
AutoCompleteBox or use a third party lib like FluentAvalonia, they offer such control

**Reply by sccombp** [reply]
AutoCompleteBox works really well with an attached behaviour. See attached dodgy example (you will need to add the Avalonia.Xaml.Behaviours nuget package to your project)

[TestAutoCompleteBehaviour.zip](https://github.com/AvaloniaUI/Avalonia/files/13037184/TestAutoCompleteBehaviour.zip)

**Reply by ManuelKetisch** [reply]
Ah lovely, the Auto-complete Box looks very promising. Thanks for pointing that out!

**Reply by ManuelKetisch** [reply]
Is there a way for the AutocompleteBox to drop open as soon as the box is selected?
In its default state you would have to know exactly what you want to type into it, in order to get autocomplete suggestions.
It would be nice to actually see the pre-defined data.

**Reply by sccombp** [reply]
If you use this behaviour 
[AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour.cs.txt](https://github.com/AvaloniaUI/Avalonia/files/13039771/AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour.cs.txt) it will display the list when you click into it (but only if it is blank)

**Reply by ManuelKetisch** [reply]
Thanks for the reply, Can you give a short rundown on how to use this class? I am fairly new to this. 

I created a class, put the code in, installed the required packages. Got no errors after building the project.

Do i need to use the standard AutocompleteBox and overwrite its default behavior somewhere/somehow?

**Reply by sccombp** [reply, marked as answer]
If you look at the example project I attached in my first reply it should give you a pretty good idea on how to use it.

Basically you add the Interactivity library and behaviour to your references at the top
```csharp
        xmlns:bh="using:TestAutoCompleteBehaviour.Behaviours"
        xmlns:i="clr-namespace:Avalonia.Xaml.Interactivity;assembly=Avalonia.Xaml.Interactivity"
```

and then on the autocomplete box you add the behaviour you are after (you can attach multiple behaviours) using the Interaction.Behaviors section to the standard AutoCompleteBox control.
```csharp
    <AutoCompleteBox Name="ACBBehaviour"
                     Margin="10"
                     MaxWidth="600"
                     IsTextCompletionEnabled="True"
                     FilterMode="Contains"
                     Watermark="Attached Behaviour (down arrow or F4 opens dropdown when empty string)">
      <i:Interaction.Behaviors>
        <bh:AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour/>
      </i:Interaction.Behaviors>
    </AutoCompleteBox>
```


full axaml code example
```csharp
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:bh="using:TestAutoCompleteBehaviour.Behaviours"
        xmlns:i="clr-namespace:Avalonia.Xaml.Interactivity;assembly=Avalonia.Xaml.Interactivity"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="TestAutoCompleteBehaviour.MainWindow"
        Title="TestAutoCompleteBehaviour">
  <Panel>
    <AutoCompleteBox Name="ACBBehaviour"
                     Margin="10"
                     MaxWidth="600"
                     IsTextCompletionEnabled="True"
                     FilterMode="Contains"
                     Watermark="Attached Behaviour (down arrow or F4 opens dropdown when empty string)">
      <i:Interaction.Behaviors>
        <bh:AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour/>
      </i:Interaction.Behaviors>
    </AutoCompleteBox>
  </Panel>
</Window>
```

**Reply by ManuelKetisch** [reply]
Ok, with your help i got it working!
Thanks for the assistance!

**Comment by jpgarza93** [comment]
Wow this works great, I made my own edit to the code above, so that:

- click opens
- typing (adding or removing) always re-evaluates and keeps it open

```cs
internal class AutoCompleteBehavior : Behavior<AutoCompleteBox>
{
    static AutoCompleteBehavior()
    {
    }

    protected override void OnAttached()
    {
        if (AssociatedObject is not null)
        {
            AssociatedObject.KeyUp += (s, e) => ShowDropdown();
            AssociatedObject.GotFocus += (s, e) => ShowDropdown();
            AssociatedObject.DropDownOpening += DropDownOpening;
        }

        base.OnAttached();
    }

    protected override void OnDetaching()
    {
        if (AssociatedObject is not null)
        {
            AssociatedObject.KeyUp += (s, e) => ShowDropdown();
            AssociatedObject.GotFocus -= (s, e) => ShowDropdown();
            AssociatedObject.DropDownOpening -= DropDownOpening;
        }

        base.OnDetaching();
    }

    private void DropDownOpening(object? sender, System.ComponentModel.CancelEventArgs e)
    {
        var prop = AssociatedObject?.GetType().GetProperty("TextBox", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic);
        var tb = (TextBox?)prop?.GetValue(AssociatedObject);
        if (tb is not null && tb.IsReadOnly)
        {
            e.Cancel = true;
            return;
        }
    }

    private void ShowDropdown()
    {
        if (AssociatedObject is not null && !AssociatedObject.IsDropDownOpen)
        {
            typeof(AutoCompleteBox).GetMethod("PopulateDropDown", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)?.Invoke(AssociatedObject, new object[] { AssociatedObject, EventArgs.Empty });
            typeof(AutoCompleteBox).GetMethod("OpeningDropDown", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)?.Invoke(AssociatedObject, new object[] { false });

            if (!AssociatedObject.IsDropDownOpen)
            {
                //We *must* set the field and not the property as we need to avoid the changed event being raised (which prevents the dropdown opening).
                var ipc = typeof(AutoCompleteBox).GetField("_ignorePropertyChange", BindingFlags.NonPublic | BindingFlags.Instance);

                if (ipc?.GetValue(AssociatedObject) is bool ipcValue && !ipcValue)
                {
                    ipc.SetValue(AssociatedObject, true);
                }

                AssociatedObject.SetCurrentValue<bool>(AutoCompleteBox.IsDropDownOpenProperty, true);
            }
        }
    }
}
```
